### ISSUE TITLE
Combine_similar can hang

### ISSUE THREAD

**Issue by mrocklin** [dask-contrib/dask-expr#343]
This happened a few times, but now I can't reproduce it .  I don't know what's going on.

```python
import dask_expr as dd
dataset_path = "s3://coiled-runtime-ci/tpch/scale-1000/"

var1 = 15
var2 = "BRASS"
var3 = "EUROPE"
region_ds = dd.read_parquet(dataset_path + "region")
nation_filtered = dd.read_parquet(dataset_path + "nation")
supplier_filtered = dd.read_parquet(dataset_path + "supplier")
part_filtered = dd.read_parquet(dataset_path + "part")
partsupp_filtered = dd.read_parquet(dataset_path + "partsupp")
region_filtered = region_ds[(region_ds["r_name"] == var3)]
r_n_merged = nation_filtered.merge(
    region_filtered, left_on="n_regionkey", right_on="r_regionkey", how="inner"
)
s_r_n_merged = r_n_merged.merge(
    supplier_filtered,
    left_on="n_nationkey",
    right_on="s_nationkey",
    how="inner",
)
ps_s_r_n_merged = s_r_n_merged.merge(
    partsupp_filtered, left_on="s_suppkey", right_on="ps_suppkey", how="inner"
)
part_filtered = part_filtered[
    (part_filtered["p_size"] == var1)
    & (part_filtered["p_type"].astype(str).str.endswith(var2))
]
merged_df = part_filtered.merge(
    ps_s_r_n_merged, left_on="p_partkey", right_on="ps_partkey", how="inner"
)
min_values = merged_df.groupby("p_partkey")["ps_supplycost"].min().reset_index()
min_values.columns = ["P_PARTKEY_CPY", "MIN_SUPPLYCOST"]
merged_df = merged_df.merge(
    min_values,
    left_on=["p_partkey", "ps_supplycost"],
    right_on=["P_PARTKEY_CPY", "MIN_SUPPLYCOST"],
    how="inner",
)
out = merged_df[
    [
        "s_acctbal",
        "s_name",
        "n_name",
        "p_partkey",
        "p_mfgr",
        "s_address",
        "s_phone",
        "s_comment",
    ]
]

out.simplify().lower_once()  # this is fine
out.combine_similar()  # this hung, but only sometimes
```

**Comment by mrocklin** [comment]
<img width="660" alt="Screenshot 2023-10-18 at 9 23 53 PM" src="https://github.com/dask-contrib/dask-expr/assets/306380/a608a521-d14b-477b-8f6f-de626020f8bb">

**Comment by mrocklin** [comment]
Yeah, we keep flopping back and forth between two states:

```(Pdb) pp expr.pprint()
BlockwiseMerge: left_on='s_suppkey' right_on='ps_suppkey'
  Projection: columns=['s_suppkey', 'n_name', 's_phone', 's_acctbal', 's_address', 's_name', 's_comment']
    DiskShuffle: partitioning_index='_partitions' npartitions_out=1746 ignore_index=False options=None
      AssignPartitioningIndex: partitioning_index=['s_suppkey'] index_name='_partitions' npartitions_out=1746
        BlockwiseMerge: left_on='n_nationkey' right_on='s_nationkey'
          Projection: columns=['n_nationkey', 'n_name']
            BlockwiseMerge: left_on='n_regionkey' right_on='r_regionkey'
              ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/nation' columns=['n_name', 'n_nationkey', 'n_regionkey'] kwargs={'dtype_backend': None}
              Filter:
                Projection: columns=['r_regionkey']
                  ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/region' columns=['r_name', 'r_regionkey'] kwargs={'dtype_backend': None}
                EQ: right='EUROPE'
                  Projection: columns='r_name'
                    ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/region' columns=['r_name', 'r_regionkey'] kwargs={'dtype_backend': None}
          ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/supplier' columns=['s_acctbal', 's_address', 's_comment', 's_name', 's_nationkey', 's_phone', 's_suppkey'] kwargs={'dtype_backend': None}
  Projection: columns=['ps_partkey', 'ps_suppkey', 'ps_supplycost']
    DiskShuffle: partitioning_index='_partitions' npartitions_out=1746 ignore_index=False options=None
      AssignPartitioningIndex: partitioning_index=['ps_suppkey'] index_name='_partitions' npartitions_out=1746
        ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/partsupp' columns=['ps_partkey', 'ps_suppkey', 'ps_supplycost'] kwargs={'dtype_backend': None}
None
(Pdb) pp out.pprint()
BlockwiseMerge: left_on='s_suppkey' right_on='ps_suppkey'
  Projection: columns=['s_suppkey', 's_name', 'n_name', 's_phone', 's_address', 's_acctbal', 's_comment']
    DiskShuffle: partitioning_index='_partitions' npartitions_out=1746 ignore_index=False options=None
      AssignPartitioningIndex: partitioning_index=['s_suppkey'] index_name='_partitions' npartitions_out=1746
        BlockwiseMerge: left_on='n_nationkey' right_on='s_nationkey'
          Projection: columns=['n_nationkey', 'n_name']
            BlockwiseMerge: left_on='n_regionkey' right_on='r_regionkey'
              ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/nation' columns=['n_name', 'n_nationkey', 'n_regionkey'] kwargs={'dtype_backend': None}
              Filter:
                Projection: columns=['r_regionkey']
                  ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/region' columns=['r_name', 'r_regionkey'] kwargs={'dtype_backend': None}
                EQ: right='EUROPE'
                  Projection: columns='r_name'
                    ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/region' columns=['r_name', 'r_regionkey'] kwargs={'dtype_backend': None}
          ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/supplier' columns=['s_acctbal', 's_address', 's_comment', 's_name', 's_nationkey', 's_phone', 's_suppkey'] kwargs={'dtype_backend': None}
  Projection: columns=['ps_partkey', 'ps_suppkey', 'ps_supplycost']
    DiskShuffle: partitioning_index='_partitions' npartitions_out=1746 ignore_index=False options=None
      AssignPartitioningIndex: partitioning_index=['ps_suppkey'] index_name='_partitions' npartitions_out=1746
        ReadParquet: path='s3://coiled-runtime-ci/tpch/scale-1000/partsupp' columns=['ps_partkey', 'ps_suppkey', 'ps_supplycost'] kwargs={'dtype_backend': None}
```

It looks like column ordering may differ.

Two concerns:

1.  Obviously there's a cycle in here somewhere
2. This algorithm isn't deterministic

I suspect that I can get around this for now by adding a `seen` dict and using that instead of `changed` for now.

**Comment by mrocklin** [comment]
@phofl here is my solution (although it's more of a bandaid than a fix for the root cause): https://github.com/dask-contrib/dask-expr/pull/349

**Comment by phofl** [comment]
The root cause if fixed in #346
